### PR TITLE
Preserve code unit sequence of directive literals

### PR DESCRIPTION
**Pull request by josephfrazier**
This is a continuation of https://github.com/prettier/prettier/pull/1560,
which was prematurely merged, then reverted in https://github.com/prettier/prettier/pull/1570.

I'm opening a new PR because GitHub doesn't let allow merged PRs to be
re-opened. You can see the unresolved discussion from the previous PR
at:

1. https://github.com/prettier/prettier/pull/1560/files#r115605758
1. https://github.com/prettier/prettier/pull/1560/files#r115606122

Fixes https://github.com/prettier/prettier/issues/1555

### CODE REVIEW

File: src/printer.js [hunk 1]
```diff
@@ -3831,21 +3847,43 @@ function nodeStr(node, options) {
...
   const enclosingQuote = shouldUseAlternateQuote
     ? alternate.quote
     : preferred.quote;
 
+  // Directives are exact code unit sequences, which means that you can't
+  // change the escape sequences they use.
+  // See https://github.com/prettier/prettier/issues/1555
+  // and https://tc39.github.io/ecma262/#directive-prologue
+  if (isDirectiveLiteral && !canChangeDirectiveQuotes) {
+    if (canChangeDirectiveQuotes) {
```
Comment (lydell) [inline src/printer.js]: `canChangeDirectiveQuotes` is always false at this point, isn't it? Should the above if statement be `if (isDirectiveLiteral) {`?

Comment (josephfrazier) [inline src/printer.js]: Thanks, fixed in https://github.com/prettier/prettier/pull/1571/commits/0b6d19db18cd64f5d191f8c01dea3ea5525503b3

File: src/printer.js [hunk 2]
```diff
@@ -181,6 +181,18 @@ function genericPrintNoParens(path, options, print, args) {
     case "EmptyStatement":
       return "";
     case "ExpressionStatement":
+      // Detect Flow-parsed directives
+      if (n.directive) {
+        return concat([
+          nodeStr(
+            n,
+            Object.assign({}, options, {
```
Comment (vjeux) [inline src/printer.js]: `options` here are the global ones we set by prettier. I was just asking for you to add a third argument when it's a flow directive literal.

File: src/printer.js [hunk 3]
```diff
@@ -181,6 +181,18 @@ function genericPrintNoParens(path, options, print, args) {
     case "EmptyStatement":
       return "";
     case "ExpressionStatement":
+      // Detect Flow-parsed directives
+      if (n.directive) {
+        return concat([
+          nodeStr(
+            n,
```
Comment (vjeux) [inline src/printer.js]: Why not pass `n.expression` here? Instead of doing all the checks inside of nodeStr?

File: src/printer.js (hunk 2 above)
Comment (josephfrazier) [inline src/printer.js]: Ah, I misunderstood. Fixed in https://github.com/prettier/prettier/pull/1571/commits/a7692e9f41a4fa324a3906f07f30016356a8f97d

File: src/printer.js (hunk 3 above)
Comment (josephfrazier) [inline src/printer.js]: Oh, that's much better. Fixed in https://github.com/prettier/prettier/pull/1571/commits/527324f5fff2cc2185fa009d244af29e4e067ffa

**Comment by vjeux** [comment]
Awesome, thanks! Sorry for so much back and forth!

**Comment by josephfrazier** [comment]
No worries, thanks for all the feedback! I'm glad we're getting closer to being able to run the AST checks in CI 👍

**Pull request merged**